### DISCUSSION TITLE
手动创建动画组件报错

### DISCUSSION THREAD

**Discussion by taohaihui** [galacean/discussion#23] [Engine] [answered]
const animator = scene1.addComponent(Animator);
        // const animator = scene1.getComponent(Animator);
        const animatorController = new AnimatorController();
        // const { animatorController } = animator;
animator.play('name') // 这里调用动画的时候会报错，
<img width="761" alt="CCC7142D-ED0C-416C-BC41-379E357BB96E" src="https://user-images.githubusercontent.com/29272705/188830400-c5a22510-dfc1-401d-9592-4a250b13b81d.png">

**Comment by singlecoder** [comment]
图片挂了，重新上传下试试？

**Comment by luzhuang** [comment, marked as answer]
const animator = scene1.addComponent(Animator);
 animator.animatorController = scene2（你动画那个模型）.getComponent(Animator).animatorController;
试试这样
模型的骨骼命名及层级要一致才可以

**Reply by luzhuang** [reply]
如果设置了动画数据，代码麻烦多截点哈

**Reply by taohaihui** [reply]
const { defaultSceneRoot: scene1, animations: animations1 = [] } = asset1;
        const { defaultSceneRoot: scene2, animations: animations2 = [] } = asset2;
        const { animations: animations3 = [] } = asset3;
        const { animations: animations4 = [] } = asset4;

        console.log(animations1, animations2, animations3, animations4);

        scene1.transform.setPosition(0, -1.2, 0);
        this.entity.addChild(scene1);

        const animator = scene1.addComponent(Animator);
        // const animator = scene1.getComponent(Animator);
        // const animatorController = new AnimatorController();
        const { animatorController } = scene2.getComponent(Animator);
        // const { animatorController } = animator;
        const additiveLayer = new AnimatorControllerLayer('layer1');
        const animatorStateMachine = new AnimatorStateMachine();
        additiveLayer.stateMachine = animatorStateMachine;
        animatorController.addLayer(additiveLayer);
        console.log('layer========>>>>', additiveLayer);

        const animationNames: string[] = [];

        if (animations2[0]) {
          const clip = animations2[0];
          // clip.name = 'welcome';
          const animatorState = animatorStateMachine.addState(clip.name);
          animatorState.clip = clip;
          animatorState.clipStartTime = 1;
          animatorState.wrapMode = WrapMode.Once;
          animationNames.push(clip.name);

          const script = animatorState.addStateMachineScript(CusStateMachineScript);
          script._onStateExit = ({ animatorState: state }) => {
            this.engine.dispatch(CustomEventName.AnimationEnd, { data: state.name });
          };
        }

        animator.animatorController = animatorController;
        console.log('animationNames======>>>>', animationNames);

        animator.play(animationNames[0] as string);

**Reply by luzhuang** [reply]
const animator = scene1.addComponent(Animator);
    const { animatorController } = scene2.getComponent(Animator);

    const animationNames: string[] = [];

    if (animations2[0]) {
      const clip = animations2[0];
      animationNames.push(clip.name);
    }

    animator.animatorController = animatorController;
    console.log('animationNames======>>>>', animationNames);

    animator.play(animationNames[0] as string);

这样试下

**Reply by luzhuang** [reply]
```
engine.resourceManager.load(["https://gw.alipayobjects.com/os/OasisHub/4c29004d-dc3d-435f-8eb5-9bb3dfdfec5c/oasis-file/1662092035431/5e3c1e4e-496e-45f8-8e05-f89f2bd5e4a4%252520%282%29.gltf", "https://gw.alipayobjects.com/os/bmw-prod/5e3c1e4e-496e-45f8-8e05-f89f2bd5e4a4.glb"])
  .then((assets) => {
    const {  defaultSceneRoot } = assets[0] as GLTFResource;
    const { defaultSceneRoot: model2, animations } = assets[1] as GLTFResource
    rootEntity.addChild(defaultSceneRoot);
    const animator = defaultSceneRoot.addComponent(Animator);
    animator.animatorController = model2.getComponent(Animator).animatorController

    animator.play(animations[0].name);
  });

```
这是我的测试代码 ok 的 可以参考下

**Reply by taohaihui** [reply]
这样是可以，因为gltf文件里面本来就有动画数据。但是现在我的是有两个gltf文件一个是模型，一个是单独的动画数据，想要把动画数据加载模型上播放的时候就会报错

**Reply by luzhuang** [reply]
你的第二个模型 动画绑定的骨骼 有层级关系么？ 这两个模型要不发我试试？ wz84888@antgroup.com

**Reply by luzhuang** [reply]
我把第一个模型的动画去掉也 ok 的
```
engine.resourceManager.load(["https://gw.alipayobjects.com/os/OasisHub/81e3860c-a77a-45fc-ac04-c099e2de9044/oasis-file/1662552341824/111.gltf", "https://gw.alipayobjects.com/os/bmw-prod/5e3c1e4e-496e-45f8-8e05-f89f2bd5e4a4.glb"])
  .then((assets) => {
    const {  defaultSceneRoot, animations: animations1 } = assets[0] as GLTFResource;
    const { defaultSceneRoot: model2, animations } = assets[1] as GLTFResource
    console.log('empty animations:', animations1)
    rootEntity.addChild(defaultSceneRoot);
    const animator = defaultSceneRoot.addComponent(Animator);
    animator.animatorController = model2.getComponent(Animator).animatorController

    animator.play(animations[0].name);
  });
```

**Reply by taohaihui** [reply]
邮件发你了

**Reply by luzhuang** [reply]
好的 我看看

**Reply by luzhuang** [reply]
hello 我看了下：
1. 原模型里把灯光也打进去了导致层级结构发生了变化，把两个模型的 defaultSceneRoot 打出来就看出来了， 层级结构要一致才可以
2. 骨骼命名不同
![image](https://user-images.githubusercontent.com/10904030/189024257-60b9df7a-2452-4fa7-b2c2-5eff8b8b5f32.png)
导致找不到正确的骨骼去绑定动画
